tensorflow_model_optimization/python/core/sparsity/keras/pruning_callbacks_test.py

```diff
@@ -55,6 +55,45 @@ def _pruned_model_setup(self, custom_training_loop=False):
       pruned_model.compile(loss=loss, optimizer=optimizer, metrics=['accuracy'])
       return pruned_model, x_train, y_train
 
+  @keras_parameterized.run_all_keras_modes
+  def testDeepLayerUpdatePruningSteps(self):
+    class TestLayer(tf.keras.layers.Layer):
+      def __init__(self):
+        super().__init__()
+        self.dense = prune.prune_low_magnitude(keras.layers.Dense(10))
+        self.head = keras.layers.Dense(5, activation='softmax')
+
+      def call(self, x):
+        x = self.dense(x)
+        return self.head(x)
+
+    class TestModel(tf.keras.Model):
+      def __init__(self):
+        super().__init__()
+        self.test_layer = TestLayer()
+
+      def call(self, x):
+        return self.test_layer(x)
+
+    x_train = np.random.rand(self._BATCH_SIZE, 10)
+    y_train = keras.utils.to_categorical(
+        np.random.randint(5, size=(self._BATCH_SIZE, 1)), 5)
+    pruned_model = TestModel()
+    pruned_model.build((self._BATCH_SIZE, 10))
+    loss = keras.losses.categorical_crossentropy
+    optimizer = keras.optimizers.SGD()
+    pruned_model.compile(loss=loss, optimizer=optimizer, metrics=['accuracy'])
+    pruned_model.fit(
+        x_train,
+        y_train,
+        batch_size=self._BATCH_SIZE,
+        epochs=3,
+        callbacks=[
+            pruning_callbacks.UpdatePruningStep()
+        ])
+
+    self.assertEqual(2, pruned_model.layers[0]._layers[0].pruning_step)
```

Comment:
For background for this PR, what's the difference between _layers and layers?

In this example, pruned_model.layers[0] is the TestLayer object.
I guess the issue is that tf.keras.layers.Layer does not have a `layers` attribute and only a `_layers` attribute?

I'm asking since ._layers is a private attribute that's not intended to be used. I can also understand the use case where TestLayer doesn't need the training functions/utilities associated with a tf.keras.Model, which does have the `layers` attribute. 

@pavithrasv: could you comment on this? We have a callback (UpdatePruningStep) that updates a variable that we add via a wrapper to every layer in a model. It's not able to access that variable inside the layers inside a layer, which would motivate this usage of a private attribute.

Reply:
@pavithrasv: ping on the above. thank you.

Reply:
Sorry about the delay. Missed the earlier msg. I don't fully remember the history behind the `layers` vs `_layers` attribute naming itself. But when you get the weights of a layers, by default it will return the weights of all the sub-layers as well. `_auto_track_sub_layers` property tracks this and unless it is explicitly turned off, layer.variables/weights should return all the sub layer weights.

If you want to get a list of sub layers including the layer itself, we have this private utility `_flatten_layers()`. `_layers` will contain Trackable containers in addition to sub-layers.

Reply:
So I should change to ```_flatten_layers(). _layers```?
I can't get the prune_step variable via ```.layers```

Reply:
Ideally calling a layer should return the sublayer weights as well. Do you know why the prune_step variable is not returned? Between `._layers` and `._flatten_layers`, `._flatten_layers` is the correct one to use.

Reply:
> Ideally calling a layer should return the sublayer weights as well. Do you know why the prune_step variable is not returned? Between `._layers` and `._flatten_layers`, `._flatten_layers` is the correct one to use.

```tuples = [(var,self.step) for var in self.model.variables if var.name.split(':')[0].split('/')[-1] == 'pruning_step']```
The above code could get pruning steps too. Should I change like this?